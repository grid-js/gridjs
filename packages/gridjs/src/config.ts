import {
  CSSDeclaration,
  OneDArray,
  ProtoExtends,
  TColumn,
  TData,
} from './types';
import Storage from './storage/storage';
import Pipeline from './pipeline/pipeline';
import Tabular from './tabular';
import { Search, SearchConfig } from './view/plugin/search/search';
import { Pagination, PaginationConfig } from './view/plugin/pagination/pagination';
import Header from './header';
import { ServerStorageOptions } from './storage/server';
import Dispatcher from './util/dispatcher';
import { GenericSortConfig } from './view/plugin/sort/sort';
import { Language, Translator } from './i18n/language';
import { Component, ComponentChild, createRef, RefObject } from 'preact';
import StorageUtils from './storage/storageUtils';
import PipelineUtils from './pipeline/pipelineUtils';
import { EventEmitter } from './util/eventEmitter';
import { GridEvents } from './events';
import { PluginManager, PluginPosition } from './plugin';
import Grid from './grid';

// Config type used internally
export interface Config {
  // a reference to the current Grid.js instance
  instance: Grid;
  eventEmitter: EventEmitter<GridEvents>;
  dispatcher: Dispatcher<any>;
  plugin: PluginManager;
  /** container element that is used to mount the Grid.js to */
  // TODO: change this to an element reference
  container?: Element;
  /** pointer to the main table element */
  tableRef?: RefObject<Component>;
  /** gridjs-temp div which is used internally */
  tempRef?: RefObject<HTMLDivElement>;
  data?: TData | (() => TData) | (() => Promise<TData>);
  server?: ServerStorageOptions;
  header?: Header;
  /** to parse a HTML table and load the data */
  from: HTMLElement;
  storage: Storage<any>;
  pipeline: Pipeline<Tabular>;
  /** to automatically calculate the columns width */
  autoWidth: boolean;
  /** sets the width of the container and table */
  width: string;
  /** sets the height of the table */
  height: string;
  pagination: PaginationConfig;
  sort: GenericSortConfig;
  translator: Translator;
  style?: Partial<{
    table: CSSDeclaration;
    td: CSSDeclaration;
    th: CSSDeclaration;
    container: CSSDeclaration;
    header: CSSDeclaration;
    footer: CSSDeclaration;
  }>;
  className?: Partial<{
    table: string;
    th: string;
    td: string;
    container: string;
    footer: string;
    header: string;
  }>;
}

// Config type used by the consumers
interface UserConfigExtend {
  /** fixes the table header to the top of the table */
  fixedHeader: boolean;
  columns: OneDArray<TColumn | string | ComponentChild>;
  search: SearchConfig | boolean;
  pagination: PaginationConfig | boolean;
  // implicit option to enable the sort plugin globally
  sort: GenericSortConfig | boolean;
  language: Language;
}

export type UserConfig = ProtoExtends<
  Partial<Config>,
  Partial<UserConfigExtend>
>;

export class Config {
  // this is the config file passed by the user
  // we need this for Config.update()
  private _userConfig: UserConfig;

  constructor(config?: Partial<Config>) {
    Object.assign(this, {
      ...Config.defaultConfig(),
      ...config,
    });

    this._userConfig = {};
  }

  /**
   * Assigns `updatedConfig` keys to the current config file
   *
   * @param updatedConfig
   */
  assign(updatedConfig: Partial<Config>): Config {
    for (const key of Object.keys(updatedConfig)) {
      // because we don't want to update the _userConfig cache
      if (key === '_userConfig') continue;

      this[key] = updatedConfig[key];
    }

    return this;
  }

  /**
   * Updates the config from a UserConfig
   *
   * @param userConfig
   */
  update(userConfig: Partial<UserConfig>): Config {
    if (!userConfig) return this;

    this._userConfig = {
      ...this._userConfig,
      ...userConfig,
    };

    this.assign(Config.fromUserConfig(this._userConfig));

    return this;
  }

  static defaultConfig(): Config {
    return {
      plugin: new PluginManager(),
      dispatcher: new Dispatcher<any>(),
      tableRef: createRef(),
      tempRef: createRef(),
      width: '100%',
      height: 'auto',
      autoWidth: true,
      style: {},
      className: {},
    } as Config;
  }

  static fromUserConfig(userConfig: UserConfig): Config {
    const config = new Config(userConfig as Config);

    // to set the initial _userConfig object
    config._userConfig = userConfig;

    // Sort
    if (typeof userConfig.sort === 'boolean' && userConfig.sort) {
      config.assign({
        sort: {
          multiColumn: true,
        },
      });
    }

    // Header
    config.assign({
      header: Header.fromUserConfig(config),
    });

    config.assign({
      storage: StorageUtils.createFromUserConfig(userConfig),
    });

    config.assign({
      pipeline: PipelineUtils.createFromConfig(config),
    });

    // Translator
    config.assign({
      translator: new Translator(userConfig.language),
    });

    // Search
    config.plugin.add({
      id: 'search',
      position: PluginPosition.Header,
      component: Search,
      props: {
        enabled:
          userConfig.search === true || userConfig.search instanceof Object,
        ...(userConfig.search as SearchConfig),
      },
    });

    // Pagination
    config.plugin.add({
      id: 'pagination',
      position: PluginPosition.Footer,
      component: Pagination,
      props: {
        type: 'opaque-based',
        enabled: userConfig.pagination === true || userConfig.pagination instanceof Object,
        ...(userConfig.pagination as PaginationConfig),
      },
    });

    return config;
  }
}
