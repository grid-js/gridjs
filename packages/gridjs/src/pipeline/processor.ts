// The order of enum items define the processing order of the processor type
// e.g. Extractor = 0 will be processed before Transformer = 1
import { generateUUID, ID } from '../util/id';
import { EventEmitter } from '../util/eventEmitter';

export enum ProcessorType {
  Initiator,
  ServerFilter,
  ServerSort,
  ServerLimit,
  Extractor,
  Transformer,
  Filter,
  Sort,
  Limit,
}

interface PipelineProcessorEvents<T, P> {
  propsUpdated: (processor: PipelineProcessor<T, P>) => void;
  beforeProcess: (...args) => void;
  afterProcess: (...args) => void;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface PipelineProcessorProps {}

export abstract class PipelineProcessor<T, P extends Partial<PipelineProcessorProps>> extends EventEmitter<PipelineProcessorEvents<T, P>> {
  public readonly id: ID;
  private readonly _props: P;

  abstract get type(): ProcessorType;
  protected abstract _process(...args): T | Promise<T>;
  protected validateProps?(...args): void;

  constructor(props?: Partial<P>) {
    super();

    this._props = {} as P;
    this.id = generateUUID();

    if (props) this.setProps(props);
  }

  /**
   * process is used to call beforeProcess and afterProcess callbacks
   * This function is just a wrapper that calls _process()
   *
   * @param args
   */
  process(...args): T | Promise<T> {
    if (this.validateProps instanceof Function) {
      this.validateProps(...args);
    }

    this.emit('beforeProcess', ...args);
    const result = this._process(...args);
    this.emit('afterProcess', ...args);
    return result;
  }

  setProps(props: Partial<P>): this {
    Object.assign(this._props, props);
    this.emit('propsUpdated', this);
    return this;
  }

  get props(): P {
    return this._props;
  }
}
