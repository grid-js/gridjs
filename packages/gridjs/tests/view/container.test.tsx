import { mount } from 'enzyme';
import { JSDOM } from 'jsdom';
import { h } from 'preact';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Config } from '../../src/config';
import { Container } from '../../src/view/container';
import StorageUtils from '../../src/storage/storageUtils';
import Header from '../../src/header';
import { PipelineProcessor, ProcessorType } from '../../src/pipeline/processor';
import * as width from '../../src/util/width';
import { flushPromises } from '../testUtil';
import PipelineUtils from '../../src/pipeline/pipelineUtils';
import { EventEmitter } from '../../src/util/eventEmitter';
import { GridEvents } from '../../src/events';
import { PluginManager } from '../../src/plugin';

expect.extend(toHaveNoViolations);

describe('Container component', () => {
  let config: Config;

  beforeEach(() => {
    config = Config.fromUserConfig({
      data: [
        [1, 2, 3],
        ['a', 'b', 'c'],
      ],
    });
    config.eventEmitter = new EventEmitter<GridEvents>();
    config.plugin = new PluginManager();
  });

  it('should render a container with table', async () => {
    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
    expect(container.state('status')).toBe(3);
  });

  it('should attach styles', async () => {
    config.update({
      search: {
        enabled: true,
      },
      pagination: {
        enabled: true,
      },
      style: {
        container: {
          border: '1px solid #ccc',
        },
        header: {
          padding: '5px',
        },
        footer: {
          margin: '2px',
        },
        td: {
          'font-weight': 'bold',
        },
        th: {
          border: '2px solid red',
        },
        table: {
          'font-size': '15px',
        },
      },
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width="500px"
        height={config.height}
      />,
    );

    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should attach classNames', async () => {
    config.update({
      search: {
        enabled: true,
      },
      pagination: {
        enabled: true,
      },
      columns: [
        {
          name: 'c1',
          sort: {
            enabled: false,
          },
        },
        {
          name: 'c2',
          sort: {
            enabled: true,
          },
        },
        {
          name: 'c3',
          sort: {
            enabled: true,
          },
        },
      ],
      className: {
        container: 'test-container',
        header: { container: 'test-header' },
        footer: { container: 'test-footer' },
        table: {
          container: 'test-table',
          td: 'test-td',
          th: 'test-th'
        },
      },
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width="500px"
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should render a container with searchable table', async () => {
    config.update({
      search: {
        enabled: true,
      },
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );
    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should render a container with sortable and paginated table', async () => {
    config.update({
      search: {
        enabled: true,
      },
      pagination: {
        enabled: true,
        limit: 5,
      },
      columns: [
        {
          name: 'c1',
          sort: {
            enabled: false,
          },
        },
        {
          name: 'c2',
          sort: {
            enabled: true,
          },
        },
        {
          name: 'c3',
          sort: {
            enabled: true,
          },
        },
      ],
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        header={config.header}
        width={config.width}
        height={config.height}
      />,
    );
    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should render a container with error', async () => {
    console.error = jest.fn();

    class ErrorProcessor extends PipelineProcessor<string, {}> {
      get type(): ProcessorType {
        return ProcessorType.Limit;
      }
      _process(): string {
        throw Error('something happened');
      }
    }

    config.pipeline.register(new ErrorProcessor());

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    await container.instance().componentDidMount();
    expect(container.html()).toMatchSnapshot();
  });

  it('should not violate accessibility test', async () => {
    config.update({
      pagination: {
        enabled: true,
        limit: 1,
      },
      search: {
        enabled: true,
      },
      sort: true,
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    await container.instance().componentDidMount();
    expect(await axe(container.html())).toHaveNoViolations();
  });

  it('should render a table with correct th widths', async () => {
    const wrapper = new JSDOM('<div style="width: 500px"></div>');

    const mock = jest.spyOn(width, 'getWidth');
    mock.mockReturnValue(42);

    config.container = wrapper;
    config.header = new Header();
    config.header.columns = [
      {
        name: 'hello world',
        sort: {
          enabled: false,
        },
      },
      {
        name: 'you are super cool',
        sort: {
          enabled: true,
        },
      },
      {
        name: 'actions',
        width: '200px',
        sort: {
          enabled: true,
        },
      },
    ];

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should render a container with null array', async () => {
    config.data = [];
    config.storage = StorageUtils.createFromUserConfig(config);
    config.pipeline = PipelineUtils.createFromConfig(config);

    config.header = new Header();
    config.header.columns = [
      {
        name: 'hello',
      },
      {
        name: 'cool',
      },
      {
        name: 'actions',
      },
    ];

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should render a container with array of objects without columns input', async () => {
    const config = Config.fromUserConfig({
      eventEmitter: new EventEmitter<GridEvents>(),
      data: [
        { name: 'boo', phoneNumber: '123' },
        { name: 'foo', phoneNumber: '456' },
        { name: 'bar', phoneNumber: '789' },
      ],
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should render a container with array of objects with string columns', async () => {
    const config = Config.fromUserConfig({
      eventEmitter: new EventEmitter<GridEvents>(),
      columns: ['Name', 'Phone Number'],
      data: [
        { name: 'boo', phoneNumber: '123' },
        { name: 'foo', phoneNumber: '456' },
        { name: 'bar', phoneNumber: '789' },
      ],
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should render a container with array of objects with object columns', async () => {
    const config = Config.fromUserConfig({
      eventEmitter: new EventEmitter<GridEvents>(),
      columns: [
        {
          name: 'Name',
          id: 'name',
        },
        {
          name: 'Phone Number',
          id: 'phone',
        },
      ],
      data: [
        { name: 'boo', phone: '123' },
        { name: 'foo', phone: '456' },
        { name: 'bar', phone: '789' },
      ],
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      expect(container.html()).toMatchSnapshot();
    });
  });

  it('should remove the EventEmitter listeners', async () => {
    config.update({
      search: {
        enabled: true,
      },
      pagination: {
        enabled: true,
      },
      columns: ['Name', 'Phone Number'],
      sort: true,
    });

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    const mockOn = jest.spyOn(EventEmitter.prototype, 'on');
    const mockOff = jest.spyOn(EventEmitter.prototype, 'off');

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      container.unmount();
      expect(mockOff.mock.calls.length).toBe(mockOn.mock.calls.length);
    });
  });

  it('should unregister the processors', async () => {
    config.update({
      pagination: {
        enabled: true,
      },
      search: {
        enabled: true,
      },
      columns: ['Name', 'Phone Number'],
      sort: true,
    });

    config.header = Header.fromUserConfig({
      columns: ['Name', 'Phone Number'],
      sort: true,
    });

    const mockRegister = jest.fn();
    const mockUnregister = jest.fn();

    config.pipeline.register = mockRegister;
    config.pipeline.unregister = mockUnregister;

    const container = mount(
      <Container
        config={config}
        pipeline={config.pipeline}
        width={config.width}
        height={config.height}
      />,
    );

    return flushPromises().then(async () => {
      await container.instance().componentDidMount();
      container.unmount();
      expect(mockUnregister.mock.calls.length).toBe(
        mockRegister.mock.calls.length,
      );
    });
  });
});
