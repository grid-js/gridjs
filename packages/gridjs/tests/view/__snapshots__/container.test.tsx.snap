// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component should attach classNames 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container test-container\\" style=\\"width: 500px;\\"><div class=\\"gridjs-head test-header\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\" style=\\"width: 500px; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table test-table\\" style=\\"width: 500px; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th data-column-id=\\"c1\\" class=\\"gridjs-th test-th\\">c1</th><th data-column-id=\\"c2\\" class=\\"gridjs-th gridjs-th-sort test-th\\" tabindex=\\"0\\">c2<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th data-column-id=\\"c3\\" class=\\"gridjs-th gridjs-th-sort test-th\\" tabindex=\\"0\\">c3<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td test-td\\" data-column-id=\\"c1\\">1</td><td class=\\"gridjs-td test-td\\" data-column-id=\\"c2\\">2</td><td class=\\"gridjs-td test-td\\" data-column-id=\\"c3\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td test-td\\" data-column-id=\\"c1\\">a</td><td class=\\"gridjs-td test-td\\" data-column-id=\\"c2\\">b</td><td class=\\"gridjs-td test-td\\" data-column-id=\\"c3\\">c</td></tr></tbody></table></div><div class=\\"gridjs-footer test-footer\\"><div class=\\"gridjs-pagination\\"><div role=\\"status\\" class=\\"gridjs-summary\\" title=\\"Page 1 of 1\\">Showing <b>1</b> to <b>2</b> of <b>2</b> results</div><div class=\\"gridjs-pages\\"><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Previous</button><button tabindex=\\"0\\" class=\\"gridjs-btn gridjs-currentPage\\" title=\\"Page 1\\">1</button><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Next</button></div></div></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should attach styles 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"border: 1px solid #ccc; width: 500px;\\"><div class=\\"gridjs-head\\" style=\\"padding: 5px;\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\" style=\\"width: 500px; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"font-size: 15px; width: 500px; height: auto;\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">1</td><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">2</td><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">a</td><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">b</td><td class=\\"gridjs-td\\" style=\\"font-weight: bold;\\">c</td></tr></tbody></table></div><div class=\\"gridjs-footer\\" style=\\"margin: 2px;\\"><div class=\\"gridjs-pagination\\"><div role=\\"status\\" class=\\"gridjs-summary\\" title=\\"Page 1 of 1\\">Showing <b>1</b> to <b>2</b> of <b>2</b> results</div><div class=\\"gridjs-pages\\"><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Previous</button><button tabindex=\\"0\\" class=\\"gridjs-btn gridjs-currentPage\\" title=\\"Page 1\\">1</button><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Next</button></div></div></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with array of objects with object columns 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th data-column-id=\\"name\\" class=\\"gridjs-th\\">Name</th><th data-column-id=\\"phone\\" class=\\"gridjs-th\\">Phone Number</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">boo</td><td class=\\"gridjs-td\\" data-column-id=\\"phone\\">123</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">foo</td><td class=\\"gridjs-td\\" data-column-id=\\"phone\\">456</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">bar</td><td class=\\"gridjs-td\\" data-column-id=\\"phone\\">789</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with array of objects with string columns 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th data-column-id=\\"name\\" class=\\"gridjs-th\\">Name</th><th data-column-id=\\"phoneNumber\\" class=\\"gridjs-th\\">Phone Number</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">boo</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">123</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">foo</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">456</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">bar</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">789</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with array of objects without columns input 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th data-column-id=\\"name\\" class=\\"gridjs-th\\">name</th><th data-column-id=\\"phoneNumber\\" class=\\"gridjs-th\\">phoneNumber</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">boo</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">123</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">foo</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">456</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\" data-column-id=\\"name\\">bar</td><td class=\\"gridjs-td\\" data-column-id=\\"phoneNumber\\">789</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with error 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td role=\\"alert\\" colspan=\\"0\\" class=\\"gridjs-td gridjs-message gridjs-error\\">An error happened while fetching the data</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with null array 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td role=\\"alert\\" colspan=\\"0\\" class=\\"gridjs-td gridjs-message gridjs-notfound\\">No matching records found</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with searchable table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-head\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with sortable and paginated table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-head\\"><div class=\\"gridjs-search\\"><input type=\\"search\\" placeholder=\\"Type a keyword...\\" aria-label=\\"Type a keyword...\\" class=\\"gridjs-input gridjs-search-input\\"></div></div><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th data-column-id=\\"c1\\" class=\\"gridjs-th\\">c1</th><th data-column-id=\\"c2\\" class=\\"gridjs-th gridjs-th-sort\\" tabindex=\\"0\\">c2<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th data-column-id=\\"c3\\" class=\\"gridjs-th gridjs-th-sort\\" tabindex=\\"0\\">c3<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td data-column-id=\\"c1\\" class=\\"gridjs-td\\">1</td><td data-column-id=\\"c2\\" class=\\"gridjs-td\\">2</td><td data-column-id=\\"c3\\" class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td data-column-id=\\"c1\\" class=\\"gridjs-td\\">a</td><td data-column-id=\\"c2\\" class=\\"gridjs-td\\">b</td><td data-column-id=\\"c3\\" class=\\"gridjs-td\\">c</td></tr></tbody></table></div><div class=\\"gridjs-footer\\"><div class=\\"gridjs-pagination\\"><div role=\\"status\\" class=\\"gridjs-summary\\" title=\\"Page 1 of 1\\">Showing <b>1</b> to <b>2</b> of <b>2</b> results</div><div class=\\"gridjs-pages\\"><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Previous</button><button tabindex=\\"0\\" class=\\"gridjs-btn gridjs-currentPage\\" title=\\"Page 1\\">1</button><button tabindex=\\"0\\" class=\\"gridjs-btn\\" disabled=\\"\\">Next</button></div></div></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a container with table 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;

exports[`Container component should render a table with correct th widths 1`] = `"<div role=\\"complementary\\" class=\\"gridjs gridjs-container\\" style=\\"width: 100%;\\"><div class=\\"gridjs-wrapper\\" style=\\"width: 100%; height: auto;\\"><table role=\\"grid\\" class=\\"gridjs-table\\" style=\\"width: 100%; height: auto;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\" style=\\"width: 42px;\\">hello world</th><th class=\\"gridjs-th gridjs-th-sort\\" style=\\"width: 42px;\\" tabindex=\\"0\\">you are super cool<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th class=\\"gridjs-th gridjs-th-sort\\" style=\\"width: 200px;\\" tabindex=\\"0\\">actions<button tabindex=\\"-1\\" aria-label=\\"Sort column ascending\\" title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table></div></div><div id=\\"gridjs-temp\\" class=\\"gridjs-temp\\"></div>"`;
